### PR TITLE
fix(treesitter): create new parser if language is not the same as cached parser

### LINKED ISSUE #18148: Tressitter: `get_parser` ignores file type change
### Neovim version (nvim -v)

0.7.0

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

Void Linux

### Terminal name/version

alacritty 0.10.1

### $TERM environment variable

alacritty

### Installation

AppImage

### How to reproduce the issue

Open a file in one file type

```sh
nvim whatever.html
```

Get the parser and inspect its language

```vim
lua p = vim.treesitter.get_parser(0)
lua print(p._lang)
```

The language is `html`, as expected. Now change the file type and get a new parser

```vim
set ft=lisp
lua q = vim.treesitter.get_parser(0)
lua print(q._lang)
```

Still prints `html`. In fact, `p == q` is `true`, they are the same parser instance.

### Expected behavior

The parser should be either different objects, or the existing parser should have its language updated. Alternatively the language of the parser should be settable via parser method.

### Actual behavior

Both parsers `p` and `q` are the same object and it has the same language as initially.

### PR DESCRIPTION
**Pull request by dchinmay2**
Currently `vim.treesitter.get_parser()` create a new parser only if there is no cached parser for that buffer. This is problematic when the filetype of a buffer is changed afterwards - the cached parser (with the wrong language) is still used.

This PR adds an additional condition for creating a new parser, ie when `cached_parser._lang` is not equal to the current lang

Closes #18148

### CODE REVIEW

**Comment by p00f** [comment]
fixes #18148

**Comment by clason** [comment]
Please use `fix(treesitter)` as scope, and describe the fix in the PR description. (The issue link need not be in the commit message.)

Would be great to have a test for this!

**Comment by clason** [comment]
An alternative would be to invalidate the cache on a filetype change:
```lua
  if parsers[bufnr] == nil then
    parsers[bufnr] = M._create_parser(bufnr, lang, opts)
    vim.api.nvim_create_autocmd("FileType", {
      buffer = bufnr,
      callback = function()
        parsers[bufnr] = nil
      end,
    })
  end
```
Might make a difference if `get_parser` is called in a hot loop somewhere (which I doubt).

**Comment by p00f** [comment]
Which 2 parsers do I use for the test?

**Comment by clason** [comment]
I'd switch from `c` (for which the parser is bundled, for this very purpose) to anything else and test that no parser is returned.

File: test/functional/treesitter/language_spec.lua [hunk 1]
```diff
@@ -67,5 +68,15 @@ describe('treesitter API', function()
     end
     eq({true,true}, {has_named,has_anonymous})
   end)
+
+  it('checks if vim.treesitter.get_parser tries to create a new parser on filetype change', function ()
+    command("set filetype=c")
+    -- Should not throw an error when filetype is c
+    eq(false, exec_lua("old_parser = vim.treesitter.get_parser(0)"))
```
Comment (clason) [inline test/functional/treesitter/language_spec.lua:75]:
```suggestion
    eq(true, exec_lua("vim.treesitter.get_parser(0)._lang == 'c'"))
```

**Comment by clason** [comment]
The highlights are using an invalid capture name; the only currently allowed captures in nvim-treesitter are those listed in https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights

**Comment by stsewd** [comment]
There is some discussion around this same problem at https://github.com/neovim/neovim/pull/14995.

File: runtime/lua/vim/treesitter.lua [hunk 2]
```diff
@@ -90,7 +90,7 @@ function M.get_parser(bufnr, lang, opts)
     lang = a.nvim_buf_get_option(bufnr, "filetype")
   end
 
-  if parsers[bufnr] == nil then
+  if parsers[bufnr] == nil or parsers[bufnr]._lang ~= lang then
```
Comment (vigoux) [inline runtime/lua/vim/treesitter.lua:93]: Please use the following here
```suggestion
  if parsers[bufnr] == nil or parsers[bufnr]:lang() ~= lang then
```

**Review by vigoux — COMMENTED** [review]
I agree with the general idea here, much more than previous attempts.

I think that @clason proposal with autocommands is trying to do too much at once.

File: test/functional/treesitter/language_spec.lua [hunk 3]
```diff
@@ -67,5 +68,15 @@ describe('treesitter API', function()
     end
     eq({true,true}, {has_named,has_anonymous})
   end)
+
+  it('checks if vim.treesitter.get_parser tries to create a new parser on filetype change', function ()
+    command("set filetype=c")
+    -- Should not throw an error when filetype is c
+    eq('c', exec_lua("return vim.treesitter.get_parser(0)._lang"))
```
Comment (clason) [inline test/functional/treesitter/language_spec.lua:75]:
```suggestion
    eq('c', exec_lua("return vim.treesitter.get_parser(0):lang()"))
```
Same thing here then, I assume?

File: runtime/lua/vim/treesitter.lua (hunk 2 above)
Comment (p00f) [reply runtime/lua/vim/treesitter.lua]: done

File: test/functional/treesitter/language_spec.lua (hunk 3 above)
Comment (p00f) [reply test/functional/treesitter/language_spec.lua]: done

**Pull request merged**